connection.go

```diff
@@ -166,12 +217,66 @@ func (connection *redisConnection) hijackConnection(name string) *redisConnectio
 	}
 }
 
+// closes a stale connection. not to be called on an active connection
+func (connection *redisConnection) closeStaleConnection() error {
+	count, err := connection.redisClient.SRem(connectionsKey, connection.Name)
+	if err != nil {
+		return err
+	}
+	if count == 0 {
+		return ErrorNotFound
+	}
+
+	// NOTE: we're not checking count here because stale connection might not
+	// have been consuming from any queue, in which case this key doesn't exist
+	if _, err = connection.redisClient.Del(connection.queuesKey); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getConsumingQueues returns a list of all queues consumed by this connection
+func (connection *redisConnection) getConsumingQueues() ([]string, error) {
+	return connection.redisClient.SMembers(connection.queuesKey)
+}
+
 // openQueue opens a queue without adding it to the set of queues
-func (connection *redisConnection) openQueue(name string) *redisQueue {
-	return newQueue(name, connection.Name, connection.queuesKey, connection.redisClient)
+func (connection *redisConnection) openQueue(name string) Queue {
+	return newQueue(
+		name,
+		connection.Name,
+		connection.queuesKey,
+		connection.redisClient,
+		connection.errChan,
+	)
+}
+
+// stopHeartbeat stops the heartbeat of the connection
+// it does not remove it from the list of connections so it can later be found by the cleaner
+func (connection *redisConnection) stopHeartbeat() error {
```

Comment:
Well spotted, you're right, there is a race here.

Unfortunately we can't just delete the method. It's used to set up test conditions. For example in cleaner tests we open a connection to set up some deliveries. Then we stop that connection's heartbeat so we can test that the cleaner properly cleans up that connection.

As tests are supposed to be quick they shouldn't actually run into the race condition in practice, but it still bothers me that the race exists. 🤔

Reply:
I think that it would be clean to move the responsibility for deleting the heartbeat key into heartbeat itself.

Then `stopHeartbeat` is only responsible for telling heartbeat to stop and waiting for it to confirm that it has stopped.

If `stopHeartbeat` performs

1. `heartbeatTicker.Stop()`
2. `close(heartbeatTocker.C)`
3. `<-heartbeatStopped`

This prevents new ticks from being sent and will terminate the `heartbeat` range loop.

`heartbeat` should probably have a deferred function which deletes the heartbeat key and closes the `heartbeatStopped` channel.

I am not sure if the `stopAllConsumers` etc. logic should also be put into the deferred function. It seems to me that if the heartbeat stops consumers should be stopped always. But I'm not sure if that breaks any tests.

Reply:
Added a commit to fix the race condition as discussed. Please have another look 👍

Reply:
Looks good.